<a href="https://colab.research.google.com/github/Amit198181/JBVNL/blob/main/Analysis_LTIS.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [6]:
import pandas as pd
import io

# Assuming the uploaded file is an Excel file
for fn in uploaded.keys():
  # Read all sheets from the Excel file into a dictionary of DataFrames
  excel_file = io.BytesIO(uploaded[fn])
  all_sheets = pd.read_excel(excel_file, sheet_name=None)
  print(f"DataFrames loaded from {fn}:")
  # Display the head of each DataFrame
  for sheet_name, df in all_sheets.items():
    print(f"\nSheet: {sheet_name}")
    display(df.head())

DataFrames loaded from FINAL_LTIS.xlsx:

Sheet: SUMMARY


Empty DataFrame
Columns: []
Index: []


Sheet: MDI>=105%CD_4_MONTHS


S.No.    ZONE  CIRCLE     DIVISION   CONSNO   ACCOUNT_NO CONSUMER_STATUS  \
0      1  RANCHI  RANCHI  RANCHI WEST  RT27482  11220279599            LIVE   
1      2  RANCHI  RANCHI  RANCHI WEST  RT28138  11220279600            LIVE   
2      3  RANCHI  RANCHI  RANCHI WEST  RT27405  11220279612            LIVE   
3      4  RANCHI  RANCHI  RANCHI WEST  RT28556  11220279628            LIVE   
4      5  RANCHI  RANCHI  RANCHI WEST  RT28602  11220279636            LIVE   

                               CONSUMER_NAME TARIFF_CODE  CONTRACTED_LOAD  \
0                M/S SABOO SONS PRATIK SABOO        LTIS             40.0   
1                               SUVESH KUMAR        LTIS              8.0   
2                               TAMIM ANSARI        LTIS              6.0   
3  SARITA SRIVASTAVA, W/O - S K SRIVASTAVA,         LTIS             10.0   
4                         SHWETA FABRICATION        LTIS             10.0   

   ... SUPPLY_HOURS_HTS_202508 CONTRACTED_LOAD_THRESHOLD  \
0  ...                       0                      42.0   
1  ...                       0                       8.4   
2  ...                       0                       6.3   
3  ...                       0                      10.5   
4  ...                       0                      10.5   

   EXCEEDED_THRESHOLD_202504 EXCEEDED_THRESHOLD_202505  \
0                      False                     False   
1                      False                     False   
2                      False                     False   
3                      False                     False   
4                      False                     False   

  EXCEEDED_THRESHOLD_202506  EXCEEDED_THRESHOLD_202507  \
0                     False                      False   
1                     False                      False   
2                     False                      False   
3                     False                      False   
4                     False                      False   

   EXCEEDED_THRESHOLD_202508  EXCEEDANCE_COUNT             TOP_4_DEMANDS  \
0                      False                 0  [18.0, 12.0, 12.0, 12.0]   
1                      False                 0      [5.0, 5.0, 5.0, 4.0]   
2                      False                 0      [6.2, 6.0, 5.0, 5.0]   
3                      False                 0      [6.0, 4.0, 3.3, 3.0]   
4                      False                 0      [6.0, 5.0, 4.4, 4.0]   

   ENHANCED_CONTRACT_DEMAND  
0                    13.500  
1                     4.750  
2                     5.550  
3                     4.075  
4                     4.850  

[5 rows x 40 columns]


Sheet: MDI>=110%CD_4_MONTHS


S.No.    ZONE  CIRCLE     DIVISION   CONSNO   ACCOUNT_NO CONSUMER_STATUS  \
0      1  RANCHI  RANCHI  RANCHI WEST  RT27349  11220279696            LIVE   
1      2  RANCHI  RANCHI  RANCHI WEST  CN11878  11220410339            LIVE   
2      3  RANCHI  RANCHI  RANCHI WEST   MC6059  11220410347            LIVE   
3      4  RANCHI  RANCHI  RANCHI WEST  CN11807  11220412226            LIVE   
4      5  RANCHI  RANCHI  RANCHI EAST   S2240N  11220541540            LIVE   

                    CONSUMER_NAME TARIFF_CODE  CONTRACTED_LOAD  ...  \
0                    THASVIEK INC        LTIS            15.00  ...   
1                  MAHATAB ANSARI        LTIS             6.00  ...   
2                 MD JUBER ANSARI        LTIS             6.00  ...   
3                   IRFAN ANSARI         LTIS             6.15  ...   
4  TARKESHWAR MANDAL S/O P.MANDAL        LTIS            13.17  ...   

  PERCENT_EXCESS_MDI_HTS_202507 PERCENT_EXCESS_MDI_HTS_202508  \
0                    120.000000                    186.666667   
1                     33.333333                     33.333333   
2                     66.666667                     33.333333   
3                     -2.439024                     30.081301   
4                     21.488231                     36.674260   

   Excess_Demand_Count Enhanced_Contract_Demand Condition_Met  \
0                    5               185.000000          True   
1                    5                33.333333          True   
2                    4                54.166667          True   
3                    4                34.146341          True   
4                    4                26.310934          True   

   PERCENT_EXCESS_MDI_HTS_202504_Met_10_Percent  \
0                                          True   
1                                          True   
2                                         False   
3                                          True   
4                                          True   

   PERCENT_EXCESS_MDI_HTS_202505_Met_10_Percent  \
0                                          True   
1                                          True   
2                                          True   
3                                          True   
4                                         False   

   PERCENT_EXCESS_MDI_HTS_202506_Met_10_Percent  \
0                                          True   
1                                          True   
2                                          True   
3                                          True   
4                                          True   

   PERCENT_EXCESS_MDI_HTS_202507_Met_10_Percent  \
0                                          True   
1                                          True   
2                                          True   
3                                         False   
4                                          True   

   PERCENT_EXCESS_MDI_HTS_202508_Met_10_Percent  
0                                          True  
1                                          True  
2                                          True  
3                                          True  
4                                          True  

[5 rows x 44 columns]


Sheet: MDI=0_ALL_MONTHS_UNITS>=100


S.No.    ZONE  CIRCLE     DIVISION  CONSNO   ACCOUNT_NO CONSUMER_STATUS  \
0      1  RANCHI  RANCHI  RANCHI WEST     GB8  11220717002            LIVE   
1      2  RANCHI  RANCHI  RANCHI EAST   M1820  11220768009            LIVE   
2      3  RANCHI  RANCHI  RANCHI EAST   M1821  11220768017            LIVE   
3      4  RANCHI  RANCHI  RANCHI EAST   S1126  11220806457            LIVE   
4      5  RANCHI  RANCHI  RANCHI EAST  OM241G  11221068343            LIVE   

                      CONSUMER_NAME TARIFF_CODE  CONTRACTED_LOAD  ...  \
0    ASSTT.ENGINEER D.W.R.S SUB-DIV        LTIS             7.02  ...   
1    ASSISTANT MECH.ENGINEER P.H.ME        LTIS            20.00  ...   
2  ASSISTANT MECH ENG.PH.MECH.SUB          LTIS            20.00  ...   
3         EXECUTIVE ENGINEER,PHED          LTIS             7.00  ...   
4        THE DY.SUPDT.ELECT.ENGG.          LTIS            55.00  ...   

  SUPPLY_HOURS_HTS_202506 BILLED_UNITS_HTS_202507  MDI_HTS_202507  \
0                       0                     404               0   
1                       0                    2938               0   
2                       0                      93               0   
3                       0                    1000               0   
4                       0                    5834               0   

  LOAD_FACTOR_HTS_202507 SUPPLY_HOURS_HTS_202507  BILLED_UNITS_HTS_202508  \
0                      0                       0                     1213   
1                      0                       0                     2693   
2                      0                       0                       85   
3                      0                       0                     1109   
4                      0                       0                     7128   

   MDI_HTS_202508  LOAD_FACTOR_HTS_202508  SUPPLY_HOURS_HTS_202508  \
0               0                       0                        0   
1               0                       0                        0   
2               0                       0                        0   
3               0                       0                        0   
4               0                       0                        0   

   100units_above  
0             YES  
1             YES  
2              NO  
3             YES  
4             YES  

[5 rows x 32 columns]


Sheet: MDI=0_LAST_MONTH_UNITS.=0


S.NO.    ZONE  CIRCLE     DIVISION  CONSNO   ACCOUNT_NO CONSUMER_STATUS  \
0      1  RANCHI  RANCHI  RANCHI WEST     GB8  11220717002            LIVE   
1      2  RANCHI  RANCHI  RANCHI EAST   M1820  11220768009            LIVE   
2      3  RANCHI  RANCHI  RANCHI EAST   S1126  11220806457            LIVE   
3      4  RANCHI  RANCHI  RANCHI EAST  TT7830  11220850561            LIVE   
4      5  RANCHI  RANCHI  RANCHI EAST  S18783  11220955749            LIVE   

                    CONSUMER_NAME TARIFF_CODE  CONTRACTED_LOAD  ...  \
0  ASSTT.ENGINEER D.W.R.S SUB-DIV        LTIS             7.02  ...   
1  ASSISTANT MECH.ENGINEER P.H.ME        LTIS            20.00  ...   
2       EXECUTIVE ENGINEER,PHED          LTIS             7.00  ...   
3     BAJRANG PRESTRESSAD PRODUCT        LTIS            88.00  ...   
4         Praveen Kumar Dumriya          LTIS            10.00  ...   

  LOAD_FACTOR_HTS_202506 SUPPLY_HOURS_HTS_202506  BILLED_UNITS_HTS_202507  \
0                    0.0                     0.0                    404.0   
1                    0.0                     0.0                   2938.0   
2                    0.0                     0.0                   1000.0   
3                    0.0                     0.0                   5936.0   
4                    0.0                     0.0                   3475.0   

  MDI_HTS_202507 LOAD_FACTOR_HTS_202507  SUPPLY_HOURS_HTS_202507  \
0            0.0                    0.0                      0.0   
1            0.0                    0.0                      0.0   
2            0.0                    0.0                      0.0   
3            0.0                    0.0                      0.0   
4            0.0                    0.0                      0.0   

   BILLED_UNITS_HTS_202508  MDI_HTS_202508  LOAD_FACTOR_HTS_202508  \
0                     1213               0                       0   
1                     2693               0                       0   
2                     1109               0                       0   
3                     3909               0                       0   
4                     3277               0                       0   

   SUPPLY_HOURS_HTS_202508  
0                        0  
1                        0  
2                        0  
3                        0  
4                        0  

[5 rows x 31 columns]


Sheet: OVER_DRAWL_wrt_CD


S.No.    ZONE  CIRCLE     DIVISION   CONSNO   ACCOUNT_NO CONSUMER_STATUS  \
0      1  RANCHI  RANCHI  RANCHI WEST  RT27482  11220279599            LIVE   
1      2  RANCHI  RANCHI  RANCHI WEST  RT28138  11220279600            LIVE   
2      3  RANCHI  RANCHI  RANCHI WEST  RT27405  11220279612            LIVE   
3      4  RANCHI  RANCHI  RANCHI WEST  RT28556  11220279628            LIVE   
4      5  RANCHI  RANCHI  RANCHI WEST  RT28602  11220279636            LIVE   

                               CONSUMER_NAME TARIFF_CODE  CONTRACTED_LOAD  \
0                M/S SABOO SONS PRATIK SABOO        LTIS             40.0   
1                               SUVESH KUMAR        LTIS              8.0   
2                               TAMIM ANSARI        LTIS              6.0   
3  SARITA SRIVASTAVA, W/O - S K SRIVASTAVA,         LTIS             10.0   
4                         SHWETA FABRICATION        LTIS             10.0   

   ... MDI_HTS_202508 LOAD_FACTOR_HTS_202508  SUPPLY_HOURS_HTS_202508  \
0  ...           18.0                    0.0                        0   
1  ...            4.0                    0.0                        0   
2  ...            5.0                    0.0                        0   
3  ...            6.0                    0.0                        0   
4  ...            6.0                    0.0                        0   

  Overdrawn Status 202504 Overdrawn Status 202505  Overdrawn Status 202506  \
0                   False                   False                    False   
1                   False                   False                    False   
2                    True                   False                    False   
3                   False                   False                    False   
4                   False                   False                    False   

   Overdrawn Status 202507  Overdrawn Status 202508  Total Overdrawn Count  \
0                    False                    False                      0   
1                    False                    False                      0   
2                    False                    False                      1   
3                    False                    False                      0   
4                    False                    False                      0   

   Over_all_result  
0            False  
1            False  
2             True  
3            False  
4            False  

[5 rows x 38 columns]

In [3]:
from google.colab import files
import os

uploaded = files.upload()

for fn in uploaded.keys():
    print('User uploaded file "{name}" with length {length} bytes'.format(
        name=fn, length=len(uploaded[fn])))

Saving FINAL_LTIS.xlsx to FINAL_LTIS.xlsx
User uploaded file "FINAL_LTIS.xlsx" with length 8329544 bytes


## Task Complete:

I have analyzed each sheet in the Excel file and presented the key findings and requested metrics for each.

Let me know if you would like to explore any of these findings further or have any other questions!

In [22]:
print(df_selected_sheet.columns)

RangeIndex(start=0, stop=0, step=1)


# Task
Analyze each sheet in the uploaded Excel file "sample_data.xlsx" separately.

## Choose a sheet

### Subtask:
Select one of the loaded sheets to analyze.


**Reasoning**:
List the keys of the all_sheets dictionary to display the names of the sheets to choose from.



In [7]:
print(all_sheets.keys())

dict_keys(['SUMMARY', 'MDI>=105%CD_4_MONTHS', 'MDI>=110%CD_4_MONTHS', 'MDI=0_ALL_MONTHS_UNITS>=100', 'MDI=0_LAST_MONTH_UNITS.=0', 'OVER_DRAWL_wrt_CD'])


**Reasoning**:
Select one of the sheets from the list for analysis and confirm the selection.



In [8]:
selected_sheet_name = 'OVER_DRAWL_wrt_CD'
print(f"Selected sheet for analysis: {selected_sheet_name}")

Selected sheet for analysis: OVER_DRAWL_wrt_CD


## Explore the data

### Subtask:
Perform initial data exploration on the selected sheet, including checking for missing values, data types, and summary statistics.


**Reasoning**:
Assign the selected sheet to a DataFrame and perform initial data exploration steps as outlined in the instructions.



In [9]:
df_selected_sheet = all_sheets[selected_sheet_name]

# Display the first 5 rows
display(df_selected_sheet.head())

# Print column names and data types
df_selected_sheet.info()

# Check for missing values
print("\nMissing values per column:")
print(df_selected_sheet.isnull().sum())

# Generate descriptive statistics for numerical columns
print("\nDescriptive statistics:")
display(df_selected_sheet.describe())

S.No.    ZONE  CIRCLE     DIVISION   CONSNO   ACCOUNT_NO CONSUMER_STATUS  \
0      1  RANCHI  RANCHI  RANCHI WEST  RT27482  11220279599            LIVE   
1      2  RANCHI  RANCHI  RANCHI WEST  RT28138  11220279600            LIVE   
2      3  RANCHI  RANCHI  RANCHI WEST  RT27405  11220279612            LIVE   
3      4  RANCHI  RANCHI  RANCHI WEST  RT28556  11220279628            LIVE   
4      5  RANCHI  RANCHI  RANCHI WEST  RT28602  11220279636            LIVE   

                               CONSUMER_NAME TARIFF_CODE  CONTRACTED_LOAD  \
0                M/S SABOO SONS PRATIK SABOO        LTIS             40.0   
1                               SUVESH KUMAR        LTIS              8.0   
2                               TAMIM ANSARI        LTIS              6.0   
3  SARITA SRIVASTAVA, W/O - S K SRIVASTAVA,         LTIS             10.0   
4                         SHWETA FABRICATION        LTIS             10.0   

   ... MDI_HTS_202508 LOAD_FACTOR_HTS_202508  SUPPLY_HOURS_HTS_202508  \
0  ...           18.0                    0.0                        0   
1  ...            4.0                    0.0                        0   
2  ...            5.0                    0.0                        0   
3  ...            6.0                    0.0                        0   
4  ...            6.0                    0.0                        0   

  Overdrawn Status 202504 Overdrawn Status 202505  Overdrawn Status 202506  \
0                   False                   False                    False   
1                   False                   False                    False   
2                    True                   False                    False   
3                   False                   False                    False   
4                   False                   False                    False   

   Overdrawn Status 202507  Overdrawn Status 202508  Total Overdrawn Count  \
0                    False                    False                      0   
1                    False                    False                      0   
2                    False                    False                      1   
3                    False                    False                      0   
4                    False                    False                      0   

   Over_all_result  
0            False  
1            False  
2             True  
3            False  
4            False  

[5 rows x 38 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17742 entries, 0 to 17741
Data columns (total 38 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   S.No.                    17742 non-null  int64  
 1   ZONE                     15715 non-null  object 
 2   CIRCLE                   15715 non-null  object 
 3   DIVISION                 15715 non-null  object 
 4   CONSNO                   15715 non-null  object 
 5   ACCOUNT_NO               17742 non-null  int64  
 6   CONSUMER_STATUS          15715 non-null  object 
 7   CONSUMER_NAME            15714 non-null  object 
 8   TARIFF_CODE              15715 non-null  object 
 9   CONTRACTED_LOAD          15715 non-null  float64
 10  CONSUMER_TYPE            15715 non-null  object 
 11  METER_READER_NAME        15703 non-null  object 
 12  MF_KWH                   15715 non-null  float64
 13  PREPAID_FLAG             15715 non-null  object 
 14  MTR_CONSUMP_TYPE      

S.No.    ACCOUNT_NO  CONTRACTED_LOAD        MF_KWH  \
count  17742.000000  1.774200e+04     15715.000000  15715.000000   
mean    8871.500000  3.999197e+10        20.072550      1.154650   
std     5121.818574  2.025792e+10        21.467967      9.710242   
min        1.000000  1.122028e+10         1.000000      1.000000   
25%     4436.250000  2.123679e+10         8.000000      1.000000   
50%     8871.500000  4.122130e+10        10.000000      1.000000   
75%    13306.750000  5.225020e+10        21.000000      1.000000   
max    17742.000000  7.225027e+10       137.000000   1200.000000   

       BILLED_UNITS_HTS_202504  MDI_HTS_202504  BILLED_UNITS_HTS_202505  \
count             1.643500e+04    16434.000000             1.612100e+04   
mean              1.943469e+03       12.600180             1.983865e+03   
std               2.208732e+04       17.803866             2.671992e+04   
min               0.000000e+00        0.000000             0.000000e+00   
25%               1.630000e+02        3.000000             1.710000e+02   
50%               4.770000e+02        7.000000             4.920000e+02   
75%               1.374000e+03       13.000000             1.400000e+03   
max               2.694078e+06      633.000000             3.235850e+06   

       MDI_HTS_202505  BILLED_UNITS_HTS_202506  MDI_HTS_202506  \
count    16121.000000             16053.000000    16053.000000   
mean        12.432317              1691.156606       12.208272   
std         16.952799              9762.403789       16.679437   
min          0.000000                 0.000000        0.000000   
25%          3.000000               161.000000        3.000000   
50%          7.000000               453.000000        7.000000   
75%         13.000000              1254.000000       12.800000   
max        270.000000            611834.000000      264.000000   

       LOAD_FACTOR_HTS_202506  SUPPLY_HOURS_HTS_202506  \
count            16053.000000                  16053.0   
mean                 0.002646                      0.0   
std                  0.179352                      0.0   
min                  0.000000                      0.0   
25%                  0.000000                      0.0   
50%                  0.000000                      0.0   
75%                  0.000000                      0.0   
max                 17.080000                      0.0   

       BILLED_UNITS_HTS_202507  MDI_HTS_202507  LOAD_FACTOR_HTS_202507  \
count             16162.000000    16162.000000            16162.000000   
mean               1582.509405       12.131869                0.001715   
std                7335.696061       16.588594                0.160460   
min                   0.000000        0.000000                0.000000   
25%                 155.000000        3.000000                0.000000   
50%                 443.000000        7.000000                0.000000   
75%                1239.000000       12.620000                0.000000   
max              504249.000000      270.000000               20.000000   

       SUPPLY_HOURS_HTS_202507  BILLED_UNITS_HTS_202508  MDI_HTS_202508  \
count                  16162.0             15715.000000    15715.000000   
mean                       0.0              1566.603945       12.139780   
std                        0.0              6638.728237       16.528773   
min                        0.0                 0.000000        0.000000   
25%                        0.0               152.000000        3.000000   
50%                        0.0               432.000000        7.000000   
75%                        0.0              1180.000000       12.840000   
max                        0.0            407237.000000      248.000000   

       LOAD_FACTOR_HTS_202508  Total Overdrawn Count  
count            15715.000000           17742.000000  
mean                 0.002910               0.779901  
std                  0.201111               1.540054  
min                  0.000000               

## Analyze the data

### Subtask:
Analyze the data in the 'OVER_DRAWL_wrt_CD' sheet to understand the patterns and characteristics of consumers who are overdrawing their contracted load.


**Reasoning**:
Analyze the distribution of 'Total Overdrawn Count', the relationship between 'CONTRACTED_LOAD' and 'Total Overdrawn Count', investigate which geographical areas have the highest occurrences of overdrawn consumers, examine the 'Over_all_result' column, and identify the correlation between 'CONTRACTED_LOAD' and 'Over_all_result'.



In [23]:
selected_sheet_name = 'OVER_DRAWL_wrt_CD'
df_selected_sheet = all_sheets[selected_sheet_name]

# 1. Analyze the distribution of 'Total Overdrawn Count'
print("Distribution of 'Total Overdrawn Count':")
print(df_selected_sheet['Total Overdrawn Count'].value_counts().sort_index())

# 2. Analyze the relationship between 'CONTRACTED_LOAD' and 'Total Overdrawn Count'
print("\nAverage 'CONTRACTED_LOAD' for each 'Total Overdrawn Count' group:")
print(df_selected_sheet.groupby('Total Overdrawn Count')['CONTRACTED_LOAD'].mean())

# 3. Investigate which Zones, Circles, and Divisions have the highest occurrences of overdrawn consumers
print("\nTop 10 Zones with highest 'Total Overdrawn Count':")
print(df_selected_sheet.groupby('ZONE')['Total Overdrawn Count'].sum().sort_values(ascending=False).head(10))

print("\nTop 10 Circles with highest 'Total Overdrawn Count':")
print(df_selected_sheet.groupby('CIRCLE')['Total Overdrawn Count'].sum().sort_values(ascending=False).head(10))

print("\nTop 10 Divisions with highest 'Total Overdrawn Count':")
print(df_selected_sheet.groupby('DIVISION')['Total Overdrawn Count'].sum().sort_values(ascending=False).head(10))

# 4. Examine the 'Over_all_result' column
print("\nPercentage of consumers with 'Over_all_result' as True:")
percentage_overall_true = (df_selected_sheet['Over_all_result'].sum() / len(df_selected_sheet)) * 100
print(f"{percentage_overall_true:.2f}%")

# 5. Identify correlation between 'CONTRACTED_LOAD' and 'Over_all_result'
print("\nMean 'CONTRACTED_LOAD' for 'Over_all_result' groups:")
print(df_selected_sheet.groupby('Over_all_result')['CONTRACTED_LOAD'].mean())

# Calculate and print the requested metrics for this sheet
print("\n--- Metrics for OVER_DRAWL_wrt_CD ---")
average_contracted_load = df_selected_sheet['CONTRACTED_LOAD'].mean()
print(f"Average contracted load: {average_contracted_load:.2f}")

contracted_load_range = f"{df_selected_sheet['CONTRACTED_LOAD'].min():.2f} - {df_selected_sheet['CONTRACTED_LOAD'].max():.2f}"
print(f"Contracted load range: {contracted_load_range}")

# Assuming MDI columns are those containing 'MDI_HTS_'
mdi_cols = [col for col in df_selected_sheet.columns if 'MDI_HTS_' in col]
average_mdi = df_selected_sheet[mdi_cols].mean().mean() # Average of averages across MDI columns
print(f"Average MDI (Maximum Demand Indicator): {average_mdi:.2f}")

# Maximum MDI value across all MDI columns and all consumers
max_mdi = df_selected_sheet[mdi_cols].max().max()
print(f"Maximum demand (MDI value): {max_mdi:.2f}")

total_consumers = len(df_selected_sheet)
print(f"Total Number of Consumers: {total_consumers}")

total_overdrawal_occurrences = df_selected_sheet['Total Overdrawn Count'].sum()
print(f"Total overdrawal occurrences (all months, all customers): {total_overdrawal_occurrences}")

Distribution of 'Total Overdrawn Count':
Total Overdrawn Count
0    13191
1     1089
2      803
3      735
4      683
5     1241
Name: count, dtype: int64

Average 'CONTRACTED_LOAD' for each 'Total Overdrawn Count' group:
Total Overdrawn Count
0    21.676399
1    16.970009
2    16.879577
3    15.711347
4    16.264173
5    15.111926
Name: CONTRACTED_LOAD, dtype: float64

Top 10 Zones with highest 'Total Overdrawn Count':
ZONE
RANCHI         2522
JAMSHEDPUR     2168
GIRIDIH        2153
HAZARIBAGH     1898
MEDININAGAR    1856
DUMKA          1658
DHANBAD        1582
Name: Total Overdrawn Count, dtype: int64

Top 10 Circles with highest 'Total Overdrawn Count':
CIRCLE
RANCHI        2153
JAMSHEDPUR    1571
DEOGHAR       1529
SAHEBGANJ     1103
DALTONGANJ    1020
KODERMA        856
GARHWA         836
DHANBAD        831
HAZARIBAGH     766
CHAS           751
Name: Total Overdrawn Count, dtype: int64

Top 10 Divisions with highest 'Total Overdrawn Count':
DIVISION
GODDA           916
RANCHI WEST

## Repeat for other sheets

### Subtask:
Go back to step 1 and repeat the process for each of the remaining sheets.


**Reasoning**:
From the `all_sheets` dictionary, select the next sheet that has not been analyzed yet, load it into a dataframe, and perform initial data exploration.



In [11]:
selected_sheet_name = 'MDI>=105%CD_4_MONTHS'
print(f"Selected sheet for analysis: {selected_sheet_name}")

df_selected_sheet = all_sheets[selected_sheet_name]

# Display the first 5 rows
display(df_selected_sheet.head())

# Print column names and data types
df_selected_sheet.info()

# Check for missing values
print("\nMissing values per column:")
print(df_selected_sheet.isnull().sum())

# Generate descriptive statistics for numerical columns
print("\nDescriptive statistics:")
display(df_selected_sheet.describe())

Selected sheet for analysis: MDI>=105%CD_4_MONTHS


S.No.    ZONE  CIRCLE     DIVISION   CONSNO   ACCOUNT_NO CONSUMER_STATUS  \
0      1  RANCHI  RANCHI  RANCHI WEST  RT27482  11220279599            LIVE   
1      2  RANCHI  RANCHI  RANCHI WEST  RT28138  11220279600            LIVE   
2      3  RANCHI  RANCHI  RANCHI WEST  RT27405  11220279612            LIVE   
3      4  RANCHI  RANCHI  RANCHI WEST  RT28556  11220279628            LIVE   
4      5  RANCHI  RANCHI  RANCHI WEST  RT28602  11220279636            LIVE   

                               CONSUMER_NAME TARIFF_CODE  CONTRACTED_LOAD  \
0                M/S SABOO SONS PRATIK SABOO        LTIS             40.0   
1                               SUVESH KUMAR        LTIS              8.0   
2                               TAMIM ANSARI        LTIS              6.0   
3  SARITA SRIVASTAVA, W/O - S K SRIVASTAVA,         LTIS             10.0   
4                         SHWETA FABRICATION        LTIS             10.0   

   ... SUPPLY_HOURS_HTS_202508 CONTRACTED_LOAD_THRESHOLD  \
0  ...                       0                      42.0   
1  ...                       0                       8.4   
2  ...                       0                       6.3   
3  ...                       0                      10.5   
4  ...                       0                      10.5   

   EXCEEDED_THRESHOLD_202504 EXCEEDED_THRESHOLD_202505  \
0                      False                     False   
1                      False                     False   
2                      False                     False   
3                      False                     False   
4                      False                     False   

  EXCEEDED_THRESHOLD_202506  EXCEEDED_THRESHOLD_202507  \
0                     False                      False   
1                     False                      False   
2                     False                      False   
3                     False                      False   
4                     False                      False   

   EXCEEDED_THRESHOLD_202508  EXCEEDANCE_COUNT             TOP_4_DEMANDS  \
0                      False                 0  [18.0, 12.0, 12.0, 12.0]   
1                      False                 0      [5.0, 5.0, 5.0, 4.0]   
2                      False                 0      [6.2, 6.0, 5.0, 5.0]   
3                      False                 0      [6.0, 4.0, 3.3, 3.0]   
4                      False                 0      [6.0, 5.0, 4.4, 4.0]   

   ENHANCED_CONTRACT_DEMAND  
0                    13.500  
1                     4.750  
2                     5.550  
3                     4.075  
4                     4.850  

[5 rows x 40 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17742 entries, 0 to 17741
Data columns (total 40 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   S.No.                      17742 non-null  int64  
 1   ZONE                       15715 non-null  object 
 2   CIRCLE                     15715 non-null  object 
 3   DIVISION                   15715 non-null  object 
 4   CONSNO                     15715 non-null  object 
 5   ACCOUNT_NO                 17742 non-null  int64  
 6   CONSUMER_STATUS            15715 non-null  object 
 7   CONSUMER_NAME              15714 non-null  object 
 8   TARIFF_CODE                15715 non-null  object 
 9   CONTRACTED_LOAD            15715 non-null  float64
 10  CONSUMER_TYPE              15715 non-null  object 
 11  METER_READER_NAME          15703 non-null  object 
 12  MF_KWH                     15715 non-null  float64
 13  PREPAID_FLAG               15715 non-null  obj

S.No.    ACCOUNT_NO  CONTRACTED_LOAD        MF_KWH  \
count  17742.000000  1.774200e+04     15715.000000  15715.000000   
mean    8871.500000  3.999197e+10        20.072550      1.154650   
std     5121.818574  2.025792e+10        21.467967      9.710242   
min        1.000000  1.122028e+10         1.000000      1.000000   
25%     4436.250000  2.123679e+10         8.000000      1.000000   
50%     8871.500000  4.122130e+10        10.000000      1.000000   
75%    13306.750000  5.225020e+10        21.000000      1.000000   
max    17742.000000  7.225027e+10       137.000000   1200.000000   

       BILLED_UNITS_HTS_202504  MDI_HTS_202504  BILLED_UNITS_HTS_202505  \
count             1.643500e+04    16434.000000             1.612100e+04   
mean              1.943469e+03       12.600180             1.983865e+03   
std               2.208732e+04       17.803866             2.671992e+04   
min               0.000000e+00        0.000000             0.000000e+00   
25%               1.630000e+02        3.000000             1.710000e+02   
50%               4.770000e+02        7.000000             4.920000e+02   
75%               1.374000e+03       13.000000             1.400000e+03   
max               2.694078e+06      633.000000             3.235850e+06   

       MDI_HTS_202505  BILLED_UNITS_HTS_202506  MDI_HTS_202506  ...  \
count    16121.000000             16053.000000    16053.000000  ...   
mean        12.432317              1691.156606       12.208272  ...   
std         16.952799              9762.403789       16.679437  ...   
min          0.000000                 0.000000        0.000000  ...   
25%          3.000000               161.000000        3.000000  ...   
50%          7.000000               453.000000        7.000000  ...   
75%         13.000000              1254.000000       12.800000  ...   
max        270.000000            611834.000000      264.000000  ...   

       BILLED_UNITS_HTS_202507  MDI_HTS_202507  LOAD_FACTOR_HTS_202507  \
count             16162.000000    16162.000000            16162.000000   
mean               1582.509405       12.131869                0.001715   
std                7335.696061       16.588594                0.160460   
min                   0.000000        0.000000                0.000000   
25%                 155.000000        3.000000                0.000000   
50%                 443.000000        7.000000                0.000000   
75%                1239.000000       12.620000                0.000000   
max              504249.000000      270.000000               20.000000   

       SUPPLY_HOURS_HTS_202507  BILLED_UNITS_HTS_202508  MDI_HTS_202508  \
count                  16162.0             15715.000000    15715.000000   
mean                       0.0              1566.603945       12.139780   
std                        0.0              6638.728237       16.528773   
min                        0.0                 0.000000        0.000000   
25%                        0.0               152.000000        3.000000   
50%                        0.0               432.000000        7.000000   
75%                        0.0              1180.000000       12.840000   
max                        0.0            407237.000000      248.000000   

       LOAD_FACTOR_HTS_202508  CONTRACTED_LOAD_THRESHOLD  EXCEEDANCE_COUNT  \
count            15715.000000               15715.000000      17742.000000   
mean                 0.002910                  21.076178          0.718803   
std                  0.201111                  22.541365          1.485061   
min                  0.000000                   1.050000          0.000000   
25%                  0.000000                   8.400000          0.000000   
50%                  0.000000                  10.500000          0.000000   
75%                  0.000000                  22.050000          0.000000   
max                 20.000000                 143.850000          5.000000   

       ENHANCED_CONTRACT_DEMAND  
count

**Reasoning**:
Analyze the data in the 'MDI>=105%CD_4_MONTHS' sheet to understand the patterns and characteristics of consumers whose MDI is at least 105% of their Contracted Load for 4 months.



In [25]:
selected_sheet_name = 'MDI>=105%CD_4_MONTHS'
df_selected_sheet = all_sheets[selected_sheet_name]

# Analyze the distribution of 'EXCEEDANCE_COUNT'
print("Distribution of 'EXCEEDANCE_COUNT':")
print(df_selected_sheet['EXCEEDANCE_COUNT'].value_counts().sort_index())

# Analyze the relationship between 'CONTRACTED_LOAD' and 'EXCEEDANCE_COUNT'
print("\nAverage 'CONTRACTED_LOAD' for each 'EXCEEDANCE_COUNT' group:")
print(df_selected_sheet.groupby('EXCEEDANCE_COUNT')['CONTRACTED_LOAD'].mean())

# Investigate which Zones, Circles, and Divisions have the highest occurrences of exceedances
print("\nTop 10 Zones with highest 'EXCEEDANCE_COUNT':")
print(df_selected_sheet.groupby('ZONE')['EXCEEDANCE_COUNT'].sum().sort_values(ascending=False).head(10))

print("\nTop 10 Circles with highest 'EXCEEDANCE_COUNT':")
print(df_selected_sheet.groupby('CIRCLE')['EXCEEDANCE_COUNT'].sum().sort_values(ascending=False).head(10))

print("\nTop 10 Divisions with highest 'EXCEEDANCE_COUNT':")
print(df_selected_sheet.groupby('DIVISION')['EXCEEDANCE_COUNT'].sum().sort_values(ascending=False).head(10))

# Analyze the distribution of 'ENHANCED_CONTRACT_DEMAND'
print("\nDescriptive statistics for 'ENHANCED_CONTRACT_DEMAND':")
display(df_selected_sheet['ENHANCED_CONTRACT_DEMAND'].describe())

# Examine the relationship between 'EXCEEDANCE_COUNT' and 'ENHANCED_CONTRACT_DEMAND'
print("\nAverage 'ENHANCED_CONTRACT_DEMAND' for each 'EXCEEDANCE_COUNT' group:")
print(df_selected_sheet.groupby('EXCEEDANCE_COUNT')['ENHANCED_CONTRACT_DEMAND'].mean())

# Calculate and print the requested metrics for this sheet
print("\n--- Metrics for MDI>=105%CD_4_MONTHS ---")
average_contracted_load = df_selected_sheet['CONTRACTED_LOAD'].mean()
print(f"Average contracted load: {average_contracted_load:.2f}")

contracted_load_range = f"{df_selected_sheet['CONTRACTED_LOAD'].min():.2f} - {df_selected_sheet['CONTRACTED_LOAD'].max():.2f}"
print(f"Contracted load range: {contracted_load_range}")

# Assuming MDI columns are those containing 'MDI_HTS_'
mdi_cols = [col for col in df_selected_sheet.columns if 'MDI_HTS_' in col]
average_mdi = df_selected_sheet[mdi_cols].mean().mean() # Average of averages across MDI columns
print(f"Average MDI (Maximum Demand Indicator): {average_mdi:.2f}")

# Maximum MDI value across all MDI columns and all consumers
max_mdi = df_selected_sheet[mdi_cols].max().max()
print(f"Maximum demand (MDI value): {max_mdi:.2f}")

total_consumers = len(df_selected_sheet)
print(f"Total Number of Consumers: {total_consumers}")

total_overdrawal_occurrences = df_selected_sheet['EXCEEDANCE_COUNT'].sum() # Using EXCEEDANCE_COUNT as the overdrawal count for this sheet
print(f"Total exceedance occurrences (all months, all customers): {total_overdrawal_occurrences}")

Distribution of 'EXCEEDANCE_COUNT':
EXCEEDANCE_COUNT
0    13487
1     1046
2      790
3      662
4      644
5     1113
Name: count, dtype: int64

Average 'CONTRACTED_LOAD' for each 'EXCEEDANCE_COUNT' group:
EXCEEDANCE_COUNT
0    21.875038
1    15.157849
2    16.061975
3    15.082674
4    15.561071
5    14.557107
Name: CONTRACTED_LOAD, dtype: float64

Top 10 Zones with highest 'EXCEEDANCE_COUNT':
ZONE
RANCHI         2207
JAMSHEDPUR     2021
GIRIDIH        2019
MEDININAGAR    1766
HAZARIBAGH     1734
DUMKA          1540
DHANBAD        1466
Name: EXCEEDANCE_COUNT, dtype: int64

Top 10 Circles with highest 'EXCEEDANCE_COUNT':
CIRCLE
RANCHI        1878
JAMSHEDPUR    1484
DEOGHAR       1465
SAHEBGANJ     1027
DALTONGANJ     961
GARHWA         805
KODERMA        782
DHANBAD        768
HAZARIBAGH     704
CHAS           698
Name: EXCEEDANCE_COUNT, dtype: int64

Top 10 Divisions with highest 'EXCEEDANCE_COUNT':
DIVISION
GODDA           895
KODERMA         782
RANCHI WEST     725
SAHEBGANJ       

count    17742.000000
mean        12.918754
std         17.043513
min          0.000000
25%          3.750000
50%          7.500000
75%         13.250000
max        264.000000
Name: ENHANCED_CONTRACT_DEMAND, dtype: float64


Average 'ENHANCED_CONTRACT_DEMAND' for each 'EXCEEDANCE_COUNT' group:
EXCEEDANCE_COUNT
0    11.121334
1    14.032134
2    16.995169
3    18.219121
4    21.103054
5    22.871412
Name: ENHANCED_CONTRACT_DEMAND, dtype: float64

--- Metrics for MDI>=105%CD_4_MONTHS ---
Average contracted load: 20.07
Contracted load range: 1.00 - 137.00
Average MDI (Maximum Demand Indicator): 12.30
Maximum demand (MDI value): 633.00
Total Number of Consumers: 17742
Total exceedance occurrences (all months, all customers): 12753


**Reasoning**:
Select the next sheet 'MDI>=110%CD_4_MONTHS' from the `all_sheets` dictionary, load it into a dataframe, and perform initial data exploration.



In [13]:
selected_sheet_name = 'MDI>=110%CD_4_MONTHS'
print(f"Selected sheet for analysis: {selected_sheet_name}")

df_selected_sheet = all_sheets[selected_sheet_name]

# Display the first 5 rows
display(df_selected_sheet.head())

# Print column names and data types
df_selected_sheet.info()

# Check for missing values
print("\nMissing values per column:")
print(df_selected_sheet.isnull().sum())

# Generate descriptive statistics for numerical columns
print("\nDescriptive statistics:")
display(df_selected_sheet.describe())

Selected sheet for analysis: MDI>=110%CD_4_MONTHS


S.No.    ZONE  CIRCLE     DIVISION   CONSNO   ACCOUNT_NO CONSUMER_STATUS  \
0      1  RANCHI  RANCHI  RANCHI WEST  RT27349  11220279696            LIVE   
1      2  RANCHI  RANCHI  RANCHI WEST  CN11878  11220410339            LIVE   
2      3  RANCHI  RANCHI  RANCHI WEST   MC6059  11220410347            LIVE   
3      4  RANCHI  RANCHI  RANCHI WEST  CN11807  11220412226            LIVE   
4      5  RANCHI  RANCHI  RANCHI EAST   S2240N  11220541540            LIVE   

                    CONSUMER_NAME TARIFF_CODE  CONTRACTED_LOAD  ...  \
0                    THASVIEK INC        LTIS            15.00  ...   
1                  MAHATAB ANSARI        LTIS             6.00  ...   
2                 MD JUBER ANSARI        LTIS             6.00  ...   
3                   IRFAN ANSARI         LTIS             6.15  ...   
4  TARKESHWAR MANDAL S/O P.MANDAL        LTIS            13.17  ...   

  PERCENT_EXCESS_MDI_HTS_202507 PERCENT_EXCESS_MDI_HTS_202508  \
0                    120.000000                    186.666667   
1                     33.333333                     33.333333   
2                     66.666667                     33.333333   
3                     -2.439024                     30.081301   
4                     21.488231                     36.674260   

   Excess_Demand_Count Enhanced_Contract_Demand Condition_Met  \
0                    5               185.000000          True   
1                    5                33.333333          True   
2                    4                54.166667          True   
3                    4                34.146341          True   
4                    4                26.310934          True   

   PERCENT_EXCESS_MDI_HTS_202504_Met_10_Percent  \
0                                          True   
1                                          True   
2                                         False   
3                                          True   
4                                          True   

   PERCENT_EXCESS_MDI_HTS_202505_Met_10_Percent  \
0                                          True   
1                                          True   
2                                          True   
3                                          True   
4                                         False   

   PERCENT_EXCESS_MDI_HTS_202506_Met_10_Percent  \
0                                          True   
1                                          True   
2                                          True   
3                                          True   
4                                          True   

   PERCENT_EXCESS_MDI_HTS_202507_Met_10_Percent  \
0                                          True   
1                                          True   
2                                          True   
3                                         False   
4                                          True   

   PERCENT_EXCESS_MDI_HTS_202508_Met_10_Percent  
0                                          True  
1                                          True  
2                                          True  
3                                          True  
4                                          True  

[5 rows x 44 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1601 entries, 0 to 1600
Data columns (total 44 columns):
 #   Column                                        Non-Null Count  Dtype  
---  ------                                        --------------  -----  
 0   S.No.                                         1601 non-null   int64  
 1   ZONE                                          1601 non-null   object 
 2   CIRCLE                                        1601 non-null   object 
 3   DIVISION                                      1601 non-null   object 
 4   CONSNO                                        1601 non-null   object 
 5   ACCOUNT_NO                                    1601 non-null   int64  
 6   CONSUMER_STATUS                               1601 non-null   object 
 7   CONSUMER_NAME                                 1601 non-null   object 
 8   TARIFF_CODE                                   1601 non-null   object 
 9   CONTRACTED_LOAD                               1601 non-null   f

S.No.    ACCOUNT_NO  CONTRACTED_LOAD      MF_KWH  \
count  1601.000000  1.601000e+03      1601.000000  1601.00000   
mean    801.000000  3.899758e+10        14.309157     1.03935   
std     462.313206  1.966570e+10        14.137153     1.09102   
min       1.000000  1.122028e+10         3.000000     1.00000   
25%     401.000000  2.123224e+10         7.020000     1.00000   
50%     801.000000  4.122157e+10         9.000000     1.00000   
75%    1201.000000  5.123705e+10        13.000000     1.00000   
max    1601.000000  7.225015e+10       107.000000    40.00000   

       BILLED_UNITS_HTS_202504  MDI_HTS_202504  BILLED_UNITS_HTS_202505  \
count              1582.000000     1582.000000              1582.000000   
mean               2767.973451       20.816422              2546.632743   
std                7599.628358       21.714222              4557.298939   
min                   0.000000        0.000000                 1.000000   
25%                 465.000000       10.000000               517.250000   
50%                1004.500000       13.000000              1069.500000   
75%                2376.500000       20.000000              2307.000000   
max              233947.000000      252.000000             63678.000000   

       MDI_HTS_202505  BILLED_UNITS_HTS_202506  MDI_HTS_202506  ...  \
count     1582.000000              1576.000000     1576.000000  ...   
mean        21.121960              2252.997462       21.264053  ...   
std         22.527329              4032.971467       22.295434  ...   
min          0.000000                 1.000000        0.000000  ...   
25%         10.000000               427.750000       10.000000  ...   
50%         13.000000               934.500000       13.000000  ...   
75%         21.000000              1973.000000       21.000000  ...   
max        270.000000             56369.000000      264.000000  ...   

       MDI_HTS_202508  LOAD_FACTOR_HTS_202508  SUPPLY_HOURS_HTS_202508  \
count     1601.000000                  1601.0                   1601.0   
mean        20.812064                     0.0                      0.0   
std         21.688975                     0.0                      0.0   
min          0.000000                     0.0                      0.0   
25%         10.000000                     0.0                      0.0   
50%         13.000000                     0.0                      0.0   
75%         20.580000                     0.0                      0.0   
max        248.000000                     0.0                      0.0   

       PERCENT_EXCESS_MDI_HTS_202504  PERCENT_EXCESS_MDI_HTS_202505  \
count                    1582.000000                    1582.000000   
mean                       46.646508                      47.634113   
std                        50.868989                      52.284317   
min                      -100.000000                    -100.000000   
25%                        19.576066                      20.000000   
50%                        33.333333                      33.333333   
75%                        60.000000                      60.000000   
max                       437.500000                     559.090909   

       PERCENT_EXCESS_MDI_HTS_202506  PERCENT_EXCESS_MDI_HTS_202507  \
count                    1576.000000                    1589.000000   
mean                       49.846650                      50.045552   
std                        53.397802                      51.595819   
min                      -100.000000                    -100.000000   
25%                        20.000000                      20.000000   
50%                        36.363636                      34.222222   
75%                        65.000000                      66.666667   
max                       481.818182                     409.090909   

       PERCENT_EXCESS_MDI_HTS_202508  Excess_Demand_Count  \
count                    1601.000000          1601.000000   
mean                       47.667761      

**Reasoning**:
Analyze the data in the 'MDI>=110%CD_4_MONTHS' sheet to understand the patterns and characteristics of consumers whose MDI is at least 110% of their Contracted Load for 4 months.



In [26]:
selected_sheet_name = 'MDI>=110%CD_4_MONTHS'
df_selected_sheet = all_sheets[selected_sheet_name]

# Analyze the distribution of 'Excess_Demand_Count'
print("Distribution of 'Excess_Demand_Count':")
print(df_selected_sheet['Excess_Demand_Count'].value_counts().sort_index())

# Analyze the relationship between 'CONTRACTED_LOAD' and 'Excess_Demand_Count'
print("\nAverage 'CONTRACTED_LOAD' for each 'Excess_Demand_Count' group:")
print(df_selected_sheet.groupby('Excess_Demand_Count')['CONTRACTED_LOAD'].mean())

# Investigate which Zones, Circles, and Divisions have the highest occurrences of excess demand
print("\nTop 10 Zones with highest 'Excess_Demand_Count':")
print(df_selected_sheet.groupby('ZONE')['Excess_Demand_Count'].sum().sort_values(ascending=False).head(10))

print("\nTop 10 Circles with highest 'Excess_Demand_Count':")
print(df_selected_sheet.groupby('CIRCLE')['Excess_Demand_Count'].sum().sort_values(ascending=False).head(10))

print("\nTop 10 Divisions with highest 'Excess_Demand_Count':")
print(df_selected_sheet.groupby('DIVISION')['Excess_Demand_Count'].sum().sort_values(ascending=False).head(10))

# Analyze the distribution of 'Enhanced_Contract_Demand'
print("\nDescriptive statistics for 'Enhanced_Contract_Demand':")
display(df_selected_sheet['Enhanced_Contract_Demand'].describe())

# Examine the relationship between 'Excess_Demand_Count' and 'Enhanced_Contract_Demand'
print("\nAverage 'Enhanced_Contract_Demand' for each 'Excess_Demand_Count' group:")
print(df_selected_sheet.groupby('Excess_Demand_Count')['Enhanced_Contract_Demand'].mean())

# Analyze the 'Condition_Met' column
print("\nPercentage of consumers with 'Condition_Met' as True:")
percentage_condition_met = (df_selected_sheet['Condition_Met'].sum() / len(df_selected_sheet)) * 100
print(f"{percentage_condition_met:.2f}%")

# Calculate and print the requested metrics for this sheet
print("\n--- Metrics for MDI>=110%CD_4_MONTHS ---")
average_contracted_load = df_selected_sheet['CONTRACTED_LOAD'].mean()
print(f"Average contracted load: {average_contracted_load:.2f}")

contracted_load_range = f"{df_selected_sheet['CONTRACTED_LOAD'].min():.2f} - {df_selected_sheet['CONTRACTED_LOAD'].max():.2f}"
print(f"Contracted load range: {contracted_load_range}")

# Assuming MDI columns are those containing 'MDI_HTS_'
mdi_cols = [col for col in df_selected_sheet.columns if 'MDI_HTS_' in col]
average_mdi = df_selected_sheet[mdi_cols].mean().mean() # Average of averages across MDI columns
print(f"Average MDI (Maximum Demand Indicator): {average_mdi:.2f}")

# Maximum MDI value across all MDI columns and all consumers
max_mdi = df_selected_sheet[mdi_cols].max().max()
print(f"Maximum demand (MDI value): {max_mdi:.2f}")

total_consumers = len(df_selected_sheet)
print(f"Total Number of Consumers: {total_consumers}")

total_overdrawal_occurrences = df_selected_sheet['Excess_Demand_Count'].sum() # Using Excess_Demand_Count as the overdrawal count for this sheet
print(f"Total excess demand occurrences (all months, all customers): {total_overdrawal_occurrences}")

Distribution of 'Excess_Demand_Count':
Excess_Demand_Count
4     580
5    1021
Name: count, dtype: int64

Average 'CONTRACTED_LOAD' for each 'Excess_Demand_Count' group:
Excess_Demand_Count
4    15.176086
5    13.816680
Name: CONTRACTED_LOAD, dtype: float64

Top 10 Zones with highest 'Excess_Demand_Count':
ZONE
RANCHI         1302
GIRIDIH        1239
JAMSHEDPUR     1140
MEDININAGAR    1048
HAZARIBAGH     1003
DUMKA           899
DHANBAD         794
Name: Excess_Demand_Count, dtype: int64

Top 10 Circles with highest 'Excess_Demand_Count':
CIRCLE
RANCHI        1125
DEOGHAR        973
JAMSHEDPUR     846
DALTONGANJ     681
SAHEBGANJ      614
KODERMA        492
DHANBAD        415
HAZARIBAGH     391
CHAS           379
GARHWA         367
Name: Excess_Demand_Count, dtype: int64

Top 10 Divisions with highest 'Excess_Demand_Count':
DIVISION
GODDA          644
KODERMA        492
SAHEBGANJ      419
DALTONGANJ     396
RANCHI WEST    381
RANCHI EAST    297
GARWAH-I       289
HAZARIBAGH     250
CHA

count    1601.000000
mean       58.683335
std        48.980544
min        11.000000
25%        27.777778
50%        42.857143
75%        74.500000
max       485.227273
Name: Enhanced_Contract_Demand, dtype: float64


Average 'Enhanced_Contract_Demand' for each 'Excess_Demand_Count' group:
Excess_Demand_Count
4    45.629941
5    66.098583
Name: Enhanced_Contract_Demand, dtype: float64

Percentage of consumers with 'Condition_Met' as True:
100.00%

--- Metrics for MDI>=110%CD_4_MONTHS ---
Average contracted load: 14.31
Contracted load range: 3.00 - 107.00
Average MDI (Maximum Demand Indicator): 23.44
Maximum demand (MDI value): 559.09
Total Number of Consumers: 1601
Total excess demand occurrences (all months, all customers): 7425


**Reasoning**:
Select the next sheet 'MDI=0_ALL_MONTHS_UNITS>=100' from the `all_sheets` dictionary, load it into a dataframe, and perform initial data exploration.



In [15]:
selected_sheet_name = 'MDI=0_ALL_MONTHS_UNITS>=100'
print(f"Selected sheet for analysis: {selected_sheet_name}")

df_selected_sheet = all_sheets[selected_sheet_name]

# Display the first 5 rows
display(df_selected_sheet.head())

# Print column names and data types
df_selected_sheet.info()

# Check for missing values
print("\nMissing values per column:")
print(df_selected_sheet.isnull().sum())

# Generate descriptive statistics for numerical columns
print("\nDescriptive statistics:")
display(df_selected_sheet.describe())

Selected sheet for analysis: MDI=0_ALL_MONTHS_UNITS>=100


S.No.    ZONE  CIRCLE     DIVISION  CONSNO   ACCOUNT_NO CONSUMER_STATUS  \
0      1  RANCHI  RANCHI  RANCHI WEST     GB8  11220717002            LIVE   
1      2  RANCHI  RANCHI  RANCHI EAST   M1820  11220768009            LIVE   
2      3  RANCHI  RANCHI  RANCHI EAST   M1821  11220768017            LIVE   
3      4  RANCHI  RANCHI  RANCHI EAST   S1126  11220806457            LIVE   
4      5  RANCHI  RANCHI  RANCHI EAST  OM241G  11221068343            LIVE   

                      CONSUMER_NAME TARIFF_CODE  CONTRACTED_LOAD  ...  \
0    ASSTT.ENGINEER D.W.R.S SUB-DIV        LTIS             7.02  ...   
1    ASSISTANT MECH.ENGINEER P.H.ME        LTIS            20.00  ...   
2  ASSISTANT MECH ENG.PH.MECH.SUB          LTIS            20.00  ...   
3         EXECUTIVE ENGINEER,PHED          LTIS             7.00  ...   
4        THE DY.SUPDT.ELECT.ENGG.          LTIS            55.00  ...   

  SUPPLY_HOURS_HTS_202506 BILLED_UNITS_HTS_202507  MDI_HTS_202507  \
0                       0                     404               0   
1                       0                    2938               0   
2                       0                      93               0   
3                       0                    1000               0   
4                       0                    5834               0   

  LOAD_FACTOR_HTS_202507 SUPPLY_HOURS_HTS_202507  BILLED_UNITS_HTS_202508  \
0                      0                       0                     1213   
1                      0                       0                     2693   
2                      0                       0                       85   
3                      0                       0                     1109   
4                      0                       0                     7128   

   MDI_HTS_202508  LOAD_FACTOR_HTS_202508  SUPPLY_HOURS_HTS_202508  \
0               0                       0                        0   
1               0                       0                        0   
2               0                       0                        0   
3               0                       0                        0   
4               0                       0                        0   

   100units_above  
0             YES  
1             YES  
2              NO  
3             YES  
4             YES  

[5 rows x 32 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 368 entries, 0 to 367
Data columns (total 32 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   S.No.                    368 non-null    int64  
 1   ZONE                     368 non-null    object 
 2   CIRCLE                   368 non-null    object 
 3   DIVISION                 368 non-null    object 
 4   CONSNO                   368 non-null    object 
 5   ACCOUNT_NO               368 non-null    int64  
 6   CONSUMER_STATUS          368 non-null    object 
 7   CONSUMER_NAME            368 non-null    object 
 8   TARIFF_CODE              368 non-null    object 
 9   CONTRACTED_LOAD          368 non-null    float64
 10  CONSUMER_TYPE            368 non-null    object 
 11  METER_READER_NAME        368 non-null    object 
 12  MF_KWH                   368 non-null    int64  
 13  PREPAID_FLAG             368 non-null    object 
 14  MTR_CONSUMP_TYPE         3

S.No.    ACCOUNT_NO  CONTRACTED_LOAD  MF_KWH  \
count  368.000000  3.680000e+02       368.000000   368.0   
mean   184.500000  4.314238e+10        19.997989     1.0   
std    106.376689  2.042608e+10        19.090652     0.0   
min      1.000000  1.122072e+10         6.000000     1.0   
25%     92.750000  3.122066e+10         8.000000     1.0   
50%    184.500000  5.122083e+10        10.000000     1.0   
75%    276.250000  6.122093e+10        25.000000     1.0   
max    368.000000  7.123398e+10       100.000000     1.0   

       BILLED_UNITS_HTS_202504  MDI_HTS_202504  BILLED_UNITS_HTS_202505  \
count               368.000000           368.0               368.000000   
mean               1811.236413             0.0              1756.660326   
std                5977.850933             0.0              3858.198268   
min                   1.000000             0.0                 1.000000   
25%                  85.000000             0.0                91.250000   
50%                 659.000000             0.0               775.000000   
75%                1690.750000             0.0              1891.750000   
max              102682.000000             0.0             57502.000000   

       MDI_HTS_202505  BILLED_UNITS_HTS_202506  MDI_HTS_202506  \
count           368.0               368.000000           368.0   
mean              0.0              1569.703804             0.0   
std               0.0              2400.960618             0.0   
min               0.0                 1.000000             0.0   
25%               0.0               102.000000             0.0   
50%               0.0               831.000000             0.0   
75%               0.0              1900.000000             0.0   
max               0.0             19587.000000             0.0   

       LOAD_FACTOR_HTS_202506  SUPPLY_HOURS_HTS_202506  \
count                   368.0                    368.0   
mean                      0.0                      0.0   
std                       0.0                      0.0   
min                       0.0                      0.0   
25%                       0.0                      0.0   
50%                       0.0                      0.0   
75%                       0.0                      0.0   
max                       0.0                      0.0   

       BILLED_UNITS_HTS_202507  MDI_HTS_202507  LOAD_FACTOR_HTS_202507  \
count               368.000000           368.0                   368.0   
mean               1615.573370             0.0                     0.0   
std                2302.762259             0.0                     0.0   
min                   1.000000             0.0                     0.0   
25%                 123.750000             0.0                     0.0   
50%                 888.000000             0.0                     0.0   
75%                1936.250000             0.0                     0.0   
max               15684.000000             0.0                     0.0   

       SUPPLY_HOURS_HTS_202507  BILLED_UNITS_HTS_202508  MDI_HTS_202508  \
count                    368.0               368.000000           368.0   
mean                       0.0              1939.127717             0.0   
std                        0.0              2746.231745             0.0   
min                        0.0                 1.000000             0.0   
25%                        0.0               187.000000             0.0   
50%                        0.0              1075.000000             0.0   
75%                        0.0              2496.000000             0.0   
max                        0.0             18966.000000             0.0   

       LOAD_FACTOR_HTS_202508  SUPPLY_HOURS_HTS_202508  
count                   368.0                    368.0  
mean                      0.0                      0.0  
std                       0.0                      0.0  
min                       0.0                      0.0  
25%                       0.0                    

**Reasoning**:
Analyze the data in the 'MDI=0_ALL_MONTHS_UNITS>=100' sheet to understand the patterns and characteristics of consumers with 0 MDI across all months but with billed units greater than or equal to 100.



In [27]:
selected_sheet_name = 'MDI=0_ALL_MONTHS_UNITS>=100'
df_selected_sheet = all_sheets[selected_sheet_name]

# Analyze the distribution of '100units_above'
print("Distribution of '100units_above':")
print(df_selected_sheet['100units_above'].value_counts())

# Analyze the relationship between 'CONTRACTED_LOAD' and '100units_above'
print("\nAverage 'CONTRACTED_LOAD' for each '100units_above' group:")
print(df_selected_sheet.groupby('100units_above')['CONTRACTED_LOAD'].mean())

# Investigate which Zones, Circles, and Divisions have the highest occurrences of '100units_above' as YES
print("\nTop 10 Zones with highest count of '100units_above' as YES:")
print(df_selected_sheet[df_selected_sheet['100units_above'] == 'YES']['ZONE'].value_counts().head(10))

print("\nTop 10 Circles with highest count of '100units_above' as YES:")
print(df_selected_sheet[df_selected_sheet['100units_above'] == 'YES']['CIRCLE'].value_counts().head(10))

print("\nTop 10 Divisions with highest count of '100units_above' as YES:")
print(df_selected_sheet[df_selected_sheet['100units_above'] == 'YES']['DIVISION'].value_counts().head(10))

# Analyze the distribution of billed units for consumers with '100units_above' as YES
print("\nDescriptive statistics for BILLED_UNITS for consumers with '100units_above' as YES:")
billed_units_cols = [col for col in df_selected_sheet.columns if 'BILLED_UNITS_HTS_' in col]
display(df_selected_sheet[df_selected_sheet['100units_above'] == 'YES'][billed_units_cols].describe())

# Calculate and print the requested metrics for this sheet
print("\n--- Metrics for MDI=0_ALL_MONTHS_UNITS>=100 ---")
average_contracted_load = df_selected_sheet['CONTRACTED_LOAD'].mean()
print(f"Average contracted load: {average_contracted_load:.2f}")

contracted_load_range = f"{df_selected_sheet['CONTRACTED_LOAD'].min():.2f} - {df_selected_sheet['CONTRACTED_LOAD'].max():.2f}"
print(f"Contracted load range: {contracted_load_range}")

# Assuming MDI columns are those containing 'MDI_HTS_'
mdi_cols = [col for col in df_selected_sheet.columns if 'MDI_HTS_' in col]
average_mdi = df_selected_sheet[mdi_cols].mean().mean() # Average of averages across MDI columns
print(f"Average MDI (Maximum Demand Indicator): {average_mdi:.2f}")

# Maximum MDI value across all MDI columns and all consumers
max_mdi = df_selected_sheet[mdi_cols].max().max()
print(f"Maximum demand (MDI value): {max_mdi:.2f}")

total_consumers = len(df_selected_sheet)
print(f"Total Number of Consumers: {total_consumers}")

# For this sheet, a relevant "overdrawal occurrence" could be the count of consumers with '100units_above' as YES
total_overdrawal_occurrences = df_selected_sheet['100units_above'].value_counts().get('YES', 0)
print(f"Total occurrences of billed units >= 100 (all months, all customers): {total_overdrawal_occurrences}")

Distribution of '100units_above':
100units_above
YES    295
NO      73
Name: count, dtype: int64

Average 'CONTRACTED_LOAD' for each '100units_above' group:
100units_above
NO     13.254795
YES    21.666644
Name: CONTRACTED_LOAD, dtype: float64

Top 10 Zones with highest count of '100units_above' as YES:
ZONE
GIRIDIH        66
RANCHI         49
DHANBAD        46
HAZARIBAGH     36
JAMSHEDPUR     36
DUMKA          35
MEDININAGAR    27
Name: count, dtype: int64

Top 10 Circles with highest count of '100units_above' as YES:
CIRCLE
RANCHI        35
GIRIDIH       33
DEOGHAR       33
CHAS          23
DHANBAD       23
JAMSHEDPUR    19
DALTONGANJ    18
SAHEBGANJ     18
DUMKA         17
CHAIBASA      17
Name: count, dtype: int64

Top 10 Divisions with highest count of '100units_above' as YES:
DIVISION
GIRIDHIH (S)    29
GODDA           17
RANCHI WEST     16
KODERMA         15
GOBINDPUR       14
PAKUR           12
SARIKELA        12
CHAS            12
DALTONGANJ      12
TENUGHAT        10
Name: co

BILLED_UNITS_HTS_202504  BILLED_UNITS_HTS_202505  \
count               295.000000               295.000000   
mean               2250.420339              2183.088136   
std                6605.422684              4202.592998   
min                   1.000000                 1.000000   
25%                 308.500000               358.500000   
50%                 881.000000              1044.000000   
75%                2152.500000              2353.000000   
max              102682.000000             57502.000000   

       BILLED_UNITS_HTS_202506  BILLED_UNITS_HTS_202507  \
count               295.000000               295.000000   
mean               1949.555932              2006.945763   
std                2542.811240              2417.515284   
min                   1.000000                 2.000000   
25%                 402.500000               645.000000   
50%                1028.000000              1102.000000   
75%                2324.000000              2448.000000   
max               19587.000000             15684.000000   

       BILLED_UNITS_HTS_202508  
count                295.00000  
mean                2410.80000  
std                 2879.03988  
min                  103.00000  
25%                  822.50000  
50%                 1267.00000  
75%                 2995.00000  
max                18966.00000


--- Metrics for MDI=0_ALL_MONTHS_UNITS>=100 ---
Average contracted load: 20.00
Contracted load range: 6.00 - 100.00
Average MDI (Maximum Demand Indicator): 0.00
Maximum demand (MDI value): 0.00
Total Number of Consumers: 368
Total occurrences of billed units >= 100 (all months, all customers): 295


**Reasoning**:
Select the next sheet 'MDI=0_LAST_MONTH_UNITS.=0' from the `all_sheets` dictionary, load it into a dataframe, and perform initial data exploration.



In [17]:
selected_sheet_name = 'MDI=0_LAST_MONTH_UNITS.=0'
print(f"Selected sheet for analysis: {selected_sheet_name}")

df_selected_sheet = all_sheets[selected_sheet_name]

# Display the first 5 rows
display(df_selected_sheet.head())

# Print column names and data types
df_selected_sheet.info()

# Check for missing values
print("\nMissing values per column:")
print(df_selected_sheet.isnull().sum())

# Generate descriptive statistics for numerical columns
print("\nDescriptive statistics:")
display(df_selected_sheet.describe())

Selected sheet for analysis: MDI=0_LAST_MONTH_UNITS.=0


S.NO.    ZONE  CIRCLE     DIVISION  CONSNO   ACCOUNT_NO CONSUMER_STATUS  \
0      1  RANCHI  RANCHI  RANCHI WEST     GB8  11220717002            LIVE   
1      2  RANCHI  RANCHI  RANCHI EAST   M1820  11220768009            LIVE   
2      3  RANCHI  RANCHI  RANCHI EAST   S1126  11220806457            LIVE   
3      4  RANCHI  RANCHI  RANCHI EAST  TT7830  11220850561            LIVE   
4      5  RANCHI  RANCHI  RANCHI EAST  S18783  11220955749            LIVE   

                    CONSUMER_NAME TARIFF_CODE  CONTRACTED_LOAD  ...  \
0  ASSTT.ENGINEER D.W.R.S SUB-DIV        LTIS             7.02  ...   
1  ASSISTANT MECH.ENGINEER P.H.ME        LTIS            20.00  ...   
2       EXECUTIVE ENGINEER,PHED          LTIS             7.00  ...   
3     BAJRANG PRESTRESSAD PRODUCT        LTIS            88.00  ...   
4         Praveen Kumar Dumriya          LTIS            10.00  ...   

  LOAD_FACTOR_HTS_202506 SUPPLY_HOURS_HTS_202506  BILLED_UNITS_HTS_202507  \
0                    0.0                     0.0                    404.0   
1                    0.0                     0.0                   2938.0   
2                    0.0                     0.0                   1000.0   
3                    0.0                     0.0                   5936.0   
4                    0.0                     0.0                   3475.0   

  MDI_HTS_202507 LOAD_FACTOR_HTS_202507  SUPPLY_HOURS_HTS_202507  \
0            0.0                    0.0                      0.0   
1            0.0                    0.0                      0.0   
2            0.0                    0.0                      0.0   
3            0.0                    0.0                      0.0   
4            0.0                    0.0                      0.0   

   BILLED_UNITS_HTS_202508  MDI_HTS_202508  LOAD_FACTOR_HTS_202508  \
0                     1213               0                       0   
1                     2693               0                       0   
2                     1109               0                       0   
3                     3909               0                       0   
4                     3277               0                       0   

   SUPPLY_HOURS_HTS_202508  
0                        0  
1                        0  
2                        0  
3                        0  
4                        0  

[5 rows x 31 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 990 entries, 0 to 989
Data columns (total 31 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   S.NO.                    990 non-null    int64  
 1   ZONE                     990 non-null    object 
 2   CIRCLE                   990 non-null    object 
 3   DIVISION                 990 non-null    object 
 4   CONSNO                   990 non-null    object 
 5   ACCOUNT_NO               990 non-null    int64  
 6   CONSUMER_STATUS          990 non-null    object 
 7   CONSUMER_NAME            990 non-null    object 
 8   TARIFF_CODE              990 non-null    object 
 9   CONTRACTED_LOAD          990 non-null    float64
 10  CONSUMER_TYPE            990 non-null    object 
 11  METER_READER_NAME        990 non-null    object 
 12  MF_KWH                   990 non-null    int64  
 13  PREPAID_FLAG             990 non-null    object 
 14  MTR_CONSUMP_TYPE         9

S.NO.    ACCOUNT_NO  CONTRACTED_LOAD      MF_KWH  \
count  990.000000  9.900000e+02       990.000000  990.000000   
mean   495.500000  4.362307e+10        22.099525    1.019192   
std    285.932684  2.001839e+10        22.074766    0.603860   
min      1.000000  1.122072e+10         2.000000    1.000000   
25%    248.250000  3.122239e+10         8.000000    1.000000   
50%    495.500000  4.123599e+10        10.000000    1.000000   
75%    742.750000  6.122092e+10        26.000000    1.000000   
max    990.000000  7.225023e+10       107.000000   20.000000   

       BILLED_UNITS_HTS_202504  MDI_HTS_202504  BILLED_UNITS_HTS_202505  \
count               918.000000      918.000000             8.950000e+02   
mean               2176.044662        7.623222             6.341947e+03   
std                8212.505684       14.466207             1.097834e+05   
min                   0.000000        0.000000             1.000000e+00   
25%                 222.250000        0.000000             2.425000e+02   
50%                 635.000000        2.000000             6.820000e+02   
75%                1673.250000        8.000000             1.795500e+03   
max              178257.000000      103.000000             3.235850e+06   

       MDI_HTS_202505  BILLED_UNITS_HTS_202506  MDI_HTS_202506  \
count      895.000000               906.000000      906.000000   
mean         6.825942              2150.564018        5.620874   
std         13.971257             13432.126725       12.287890   
min          0.000000                 1.000000        0.000000   
25%          0.000000               237.250000        0.000000   
50%          1.000000               630.500000        0.000000   
75%          7.000000              1645.750000        6.000000   
max        114.000000            393292.000000       95.000000   

       LOAD_FACTOR_HTS_202506  SUPPLY_HOURS_HTS_202506  \
count                   906.0                    906.0   
mean                      0.0                      0.0   
std                       0.0                      0.0   
min                       0.0                      0.0   
25%                       0.0                      0.0   
50%                       0.0                      0.0   
75%                       0.0                      0.0   
max                       0.0                      0.0   

       BILLED_UNITS_HTS_202507  MDI_HTS_202507  LOAD_FACTOR_HTS_202507  \
count               926.000000      926.000000                   926.0   
mean               1936.917927        4.212527                     0.0   
std                7454.351558       10.425834                     0.0   
min                   1.000000        0.000000                     0.0   
25%                 239.000000        0.000000                     0.0   
50%                 694.500000        0.000000                     0.0   
75%                1750.250000        4.315000                     0.0   
max              203628.000000      109.000000                     0.0   

       SUPPLY_HOURS_HTS_202507  BILLED_UNITS_HTS_202508  MDI_HTS_202508  \
count                    926.0               990.000000           990.0   
mean                       0.0              2406.583838             0.0   
std                        0.0              9946.291281             0.0   
min                        0.0               100.000000             0.0   
25%                        0.0               260.000000             0.0   
50%                        0.0               763.000000             0.0   
75%                        0.0              1797.500000             0.0   
max                        0.0            207272.000000             0.0   

       LOAD_FACTOR_HTS_202508  SUPPLY_HOURS_HTS_202508  
count                   990.0                    990.0  
mean                      0.0                      0.0  
std                       0.0                      0.0  
min                       0.0                      0.0  
25%          

**Reasoning**:
Analyze the data in the 'MDI=0_LAST_MONTH_UNITS.=0' sheet to understand the patterns and characteristics of consumers with 0 MDI and 0 billed units in the last month.



In [28]:
selected_sheet_name = 'MDI=0_LAST_MONTH_UNITS.=0'
df_selected_sheet = all_sheets[selected_sheet_name]

# Analyze the distribution of billed units for the last month ('BILLED_UNITS_HTS_202508')
print("Distribution of 'BILLED_UNITS_HTS_202508':")
print(df_selected_sheet['BILLED_UNITS_HTS_202508'].value_counts().sort_index().head())

# Analyze the distribution of MDI for the last month ('MDI_HTS_202508')
print("\nDistribution of 'MDI_HTS_202508':")
print(df_selected_sheet['MDI_HTS_202508'].value_counts())

# Analyze the relationship between 'CONTRACTED_LOAD' and the last month's billed units
print("\nAverage 'CONTRACTED_LOAD' for different ranges of 'BILLED_UNITS_HTS_202508':")
bins = [0, 100, 500, 1000, 5000, df_selected_sheet['BILLED_UNITS_HTS_202508'].max()]
labels = ['0-100', '101-500', '501-1000', '1001-5000', '>5000']
df_selected_sheet['BILLED_UNITS_BIN'] = pd.cut(df_selected_sheet['BILLED_UNITS_HTS_202508'], bins=bins, labels=labels, right=False)
print(df_selected_sheet.groupby('BILLED_UNITS_BIN')['CONTRACTED_LOAD'].mean())

# Investigate which Zones, Circles, and Divisions have the highest occurrences of 0 billed units in the last month
# Assuming 0 billed units is a key characteristic of this sheet based on the name.
# Need to confirm this from the data description or by checking value counts for the last month billed units.
# Since value counts show min is 100, the sheet name might be slightly misleading or refers to another column.
# Re-examining the data info, BILLED_UNITS_HTS_202508 has a min of 100.
# Let's analyze the distribution of billed units and MDI for the last month to understand the data.

print("\nTop 10 Zones with lowest average 'BILLED_UNITS_HTS_202508':")
print(df_selected_sheet.groupby('ZONE')['BILLED_UNITS_HTS_202508'].mean().sort_values().head(10))

print("\nTop 10 Circles with lowest average 'BILLED_UNITS_HTS_202508':")
print(df_selected_sheet.groupby('CIRCLE')['BILLED_UNITS_HTS_202508'].mean().sort_values().head(10))

print("\nTop 10 Divisions with lowest average 'BILLED_UNITS_HTS_202508':")
print(df_selected_sheet.groupby('DIVISION')['BILLED_UNITS_HTS_202508'].mean().sort_values().head(10))

# Calculate and print the requested metrics for this sheet
print("\n--- Metrics for MDI=0_LAST_MONTH_UNITS.=0 ---")
average_contracted_load = df_selected_sheet['CONTRACTED_LOAD'].mean()
print(f"Average contracted load: {average_contracted_load:.2f}")

contracted_load_range = f"{df_selected_sheet['CONTRACTED_LOAD'].min():.2f} - {df_selected_sheet['CONTRACTED_LOAD'].max():.2f}"
print(f"Contracted load range: {contracted_load_range}")

# Assuming MDI columns are those containing 'MDI_HTS_'
mdi_cols = [col for col in df_selected_sheet.columns if 'MDI_HTS_' in col]
average_mdi = df_selected_sheet[mdi_cols].mean().mean() # Average of averages across MDI columns
print(f"Average MDI (Maximum Demand Indicator): {average_mdi:.2f}")

# Maximum MDI value across all MDI columns and all consumers
max_mdi = df_selected_sheet[mdi_cols].max().max()
print(f"Maximum demand (MDI value): {max_mdi:.2f}")

total_consumers = len(df_selected_sheet)
print(f"Total Number of Consumers: {total_consumers}")

# For this sheet, a relevant metric could be the number of consumers with 0 MDI in the last month
total_overdrawal_occurrences = df_selected_sheet[df_selected_sheet['MDI_HTS_202508'] == 0].shape[0]
print(f"Total consumers with 0 MDI in the last month: {total_overdrawal_occurrences}")

Distribution of 'BILLED_UNITS_HTS_202508':
BILLED_UNITS_HTS_202508
100    2
101    3
102    1
103    3
104    4
Name: count, dtype: int64

Distribution of 'MDI_HTS_202508':
MDI_HTS_202508
0    990
Name: count, dtype: int64

Average 'CONTRACTED_LOAD' for different ranges of 'BILLED_UNITS_HTS_202508':
BILLED_UNITS_BIN
0-100              NaN
101-500      14.563184
501-1000     16.090402
1001-5000    25.820500
>5000        53.159570
Name: CONTRACTED_LOAD, dtype: float64

Top 10 Zones with lowest average 'BILLED_UNITS_HTS_202508':
ZONE
MEDININAGAR    1415.920635
GIRIDIH        1582.859375
DUMKA          1657.361111
DHANBAD        1941.368984
JAMSHEDPUR     2628.684211
RANCHI         3008.312057
HAZARIBAGH     4805.073394
Name: BILLED_UNITS_HTS_202508, dtype: float64

Top 10 Circles with lowest average 'BILLED_UNITS_HTS_202508':
CIRCLE
GARHWA         985.733333
GIRIDIH       1540.092784
RANCHI        1543.918605
SAHEBGANJ     1545.977273
DALTONGANJ    1550.354167
DEOGHAR       1626.526316
DU

/tmp/ipython-input-2209898162.py:17: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  print(df_selected_sheet.groupby('BILLED_UNITS_BIN')['CONTRACTED_LOAD'].mean())


## Repeat for other sheets

### Subtask:
Go back to step 1 and repeat the process for each of the remaining sheets.


**Reasoning**:
Select the next unanalyzed sheet and perform the initial data exploration steps: load the sheet, display head, info, and missing values.



In [19]:
selected_sheet_name = 'SUMMARY'
print(f"Selected sheet for analysis: {selected_sheet_name}")

df_selected_sheet = all_sheets[selected_sheet_name]

# Display the first 5 rows
display(df_selected_sheet.head())

# Print column names and data types
df_selected_sheet.info()

# Check for missing values
print("\nMissing values per column:")
print(df_selected_sheet.isnull().sum())

# Generate descriptive statistics for numerical columns
print("\nDescriptive statistics:")
display(df_selected_sheet.describe())

Selected sheet for analysis: SUMMARY


Empty DataFrame
Columns: []
Index: []

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 0 entries
Empty DataFrame

Missing values per column:
Series([], dtype: float64)

Descriptive statistics:


ValueError: Cannot describe a DataFrame without columns

## Summarize findings

### Subtask:
Summarize the key findings from the analysis of each sheet.


**Reasoning**:
Review the analysis results from each sheet and synthesize the key findings, noting that the 'SUMMARY' sheet was empty.



In [29]:
print("Summary of Key Findings from Each Sheet Analysis:")
print("\n--- OVER_DRAWL_wrt_CD ---")
print("- Distribution of Total Overdrawn Count: Majority of consumers (13,191) had 0 overdrawn counts. Counts range from 1 to 5 for those who overdraw.")
print("- Relationship between Contracted Load and Total Overdrawn Count: Consumers who overdraw tend to have a lower average Contracted Load compared to those who don't.")
print("- Geographical Hotspots for Overdrawing: RANCHI, JAMSHEDPUR, and GIRIDIH zones; RANCHI, JAMSHEDPUR, and DEOGHAR circles; GODDA, RANCHI WEST, and KODERMA divisions showed the highest total overdrawn counts.")
print(f"- Overall Overdrawing Proportion: {percentage_overall_true:.2f}% of consumers identified as overdrawing.")
print(f"- Average contracted load: {all_sheets['OVER_DRAWL_wrt_CD']['CONTRACTED_LOAD'].mean():.2f}")
print(f"- Contracted load range: {all_sheets['OVER_DRAWL_wrt_CD']['CONTRACTED_LOAD'].min():.2f} - {all_sheets['OVER_DRAWL_wrt_CD']['CONTRACTED_LOAD'].max():.2f}")
mdi_cols_overdrawl = [col for col in all_sheets['OVER_DRAWL_wrt_CD'].columns if 'MDI_HTS_' in col]
print(f"- Average MDI (Maximum Demand Indicator): {all_sheets['OVER_DRAWL_wrt_CD'][mdi_cols_overdrawl].mean().mean():.2f}")
print(f"- Maximum demand (MDI value): {all_sheets['OVER_DRAWL_wrt_CD'][mdi_cols_overdrawl].max().max():.2f}")
print(f"- Total Number of Consumers: {len(all_sheets['OVER_DRAWL_wrt_CD'])}")
print(f"- Total overdrawal occurrences (all months, all customers): {all_sheets['OVER_DRAWL_wrt_CD']['Total Overdrawn Count'].sum()}")


print("\n--- MDI>=105%CD_4_MONTHS ---")
print("- Distribution of EXCEEDANCE_COUNT: Many consumers had 0 exceedances, but a portion had 1 to 5 exceedances.")
print("- Relationship between Contracted Load and EXCEEDANCE_COUNT: Higher exceedance counts were associated with slightly lower average Contracted Load but higher average Enhanced Contract Demand.")
print("- Geographical Hotspots for Exceedances: Top Zones, Circles, and Divisions with highest sum of EXCEEDANCE_COUNT were identified.")
print(f"- Average contracted load: {all_sheets['MDI>=105%CD_4_MONTHS']['CONTRACTED_LOAD'].mean():.2f}")
print(f"- Contracted load range: {all_sheets['MDI>=105%CD_4_MONTHS']['CONTRACTED_LOAD'].min():.2f} - {all_sheets['MDI>=105%CD_4_MONTHS']['CONTRACTED_LOAD'].max():.2f}")
mdi_cols_105 = [col for col in all_sheets['MDI>=105%CD_4_MONTHS'].columns if 'MDI_HTS_' in col]
print(f"- Average MDI (Maximum Demand Indicator): {all_sheets['MDI>=105%CD_4_MONTHS'][mdi_cols_105].mean().mean():.2f}")
print(f"- Maximum demand (MDI value): {all_sheets['MDI>=105%CD_4_MONTHS'][mdi_cols_105].max().max():.2f}")
print(f"- Total Number of Consumers: {len(all_sheets['MDI>=105%CD_4_MONTHS'])}")
print(f"- Total exceedance occurrences (all months, all customers): {all_sheets['MDI>=105%CD_4_MONTHS']['EXCEEDANCE_COUNT'].sum()}")


print("\n--- MDI>=110%CD_4_MONTHS ---")
print("- Distribution of Excess_Demand_Count: All consumers in this sheet had an excess demand count of 4 or 5.")
print("- Relationship between Contracted Load and Excess_Demand_Count: Consumers with a count of 5 had slightly lower average Contracted Load but higher average Enhanced Contract Demand.")
print("- Geographical Hotspots for Excess Demand: Top geographical areas with the highest sum of Excess_Demand_Count were identified.")
print(f"- Condition Met: {percentage_condition_met:.2f}% of consumers in this sheet met the specified condition.")
print(f"- Average contracted load: {all_sheets['MDI>=110%CD_4_MONTHS']['CONTRACTED_LOAD'].mean():.2f}")
print(f"- Contracted load range: {all_sheets['MDI>=110%CD_4_MONTHS']['CONTRACTED_LOAD'].min():.2f} - {all_sheets['MDI>=110%CD_4_MONTHS']['CONTRACTED_LOAD'].max():.2f}")
mdi_cols_110 = [col for col in all_sheets['MDI>=110%CD_4_MONTHS'].columns if 'MDI_HTS_' in col]
print(f"- Average MDI (Maximum Demand Indicator): {all_sheets['MDI>=110%CD_4_MONTHS'][mdi_cols_110].mean().mean():.2f}")
print(f"- Maximum demand (MDI value): {all_sheets['MDI>=110%CD_4_MONTHS'][mdi_cols_110].max().max():.2f}")
print(f"- Total Number of Consumers: {len(all_sheets['MDI>=110%CD_4_MONTHS'])}")
print(f"- Total excess demand occurrences (all months, all customers): {all_sheets['MDI>=110%CD_4_MONTHS']['Excess_Demand_Count'].sum()}")


print("\n--- MDI=0_ALL_MONTHS_UNITS>=100 ---")
print("- Distribution of 100units_above: Majority of consumers had billed units above 100 ('YES').")
print("- Relationship between Contracted Load and 100units_above: Consumers with 'YES' had a significantly higher average Contracted Load.")
print("- Geographical Hotspots for '100units_above' as YES: Top geographical areas were identified.")
print("- Billed Units: Varied ranges and averages of consumption observed for consumers with '100units_above' as 'YES'.")
print(f"- Average contracted load: {all_sheets['MDI=0_ALL_MONTHS_UNITS>=100']['CONTRACTED_LOAD'].mean():.2f}")
print(f"- Contracted load range: {all_sheets['MDI=0_ALL_MONTHS_UNITS>=100']['CONTRACTED_LOAD'].min():.2f} - {all_sheets['MDI=0_ALL_MONTHS_UNITS>=100']['CONTRACTED_LOAD'].max():.2f}")
mdi_cols_0_100 = [col for col in all_sheets['MDI=0_ALL_MONTHS_UNITS>=100'].columns if 'MDI_HTS_' in col]
print(f"- Average MDI (Maximum Demand Indicator): {all_sheets['MDI=0_ALL_MONTHS_UNITS>=100'][mdi_cols_0_100].mean().mean():.2f}")
print(f"- Maximum demand (MDI value): {all_sheets['MDI=0_ALL_MONTHS_UNITS>=100'][mdi_cols_0_100].max().max():.2f}")
print(f"- Total Number of Consumers: {len(all_sheets['MDI=0_ALL_MONTHS_UNITS>=100'])}")
print(f"- Total occurrences of billed units >= 100 (all months, all customers): {all_sheets['MDI=0_ALL_MONTHS_UNITS>=100']['100units_above'].value_counts().get('YES', 0)}")


print("\n--- MDI=0_LAST_MONTH_UNITS.=0 ---")
print("- MDI in last month: All consumers had an MDI of 0 in the last month.")
print("- Billed Units in last month: Minimum billed units in the last month was 100, contradicting the sheet name.")
print("- Relationship between Contracted Load and last month's billed units: Positive correlation observed.")
print("- Geographical Areas with lowest average billed units: Identified top geographical areas with the lowest average billed units in the last month.")
print(f"- Average contracted load: {all_sheets['MDI=0_LAST_MONTH_UNITS.=0']['CONTRACTED_LOAD'].mean():.2f}")
print(f"- Contracted load range: {all_sheets['MDI=0_LAST_MONTH_UNITS.=0']['CONTRACTED_LOAD'].min():.2f} - {all_sheets['MDI=0_LAST_MONTH_UNITS.=0']['CONTRACTED_LOAD'].max():.2f}")
mdi_cols_0_0 = [col for col in all_sheets['MDI=0_LAST_MONTH_UNITS.=0'].columns if 'MDI_HTS_' in col]
print(f"- Average MDI (Maximum Demand Indicator): {all_sheets['MDI=0_LAST_MONTH_UNITS.=0'][mdi_cols_0_0].mean().mean():.2f}")
print(f"- Maximum demand (MDI value): {all_sheets['MDI=0_LAST_MONTH_UNITS.=0'][mdi_cols_0_0].max().max():.2f}")
print(f"- Total Number of Consumers: {len(all_sheets['MDI=0_LAST_MONTH_UNITS.=0'])}")
print(f"- Total consumers with 0 MDI in the last month: {all_sheets['MDI=0_LAST_MONTH_UNITS.=0'][all_sheets['MDI=0_LAST_MONTH_UNITS.=0']['MDI_HTS_202508'] == 0].shape[0]}")


print("\n--- SUMMARY ---")
if not all_sheets['SUMMARY'].empty:
    print("- The 'SUMMARY' sheet contained data.")
    # Add specific analysis for the SUMMARY sheet if needed
else:
    print("- The 'SUMMARY' sheet was found to be empty and contained no data for analysis.")

Summary of Key Findings from Each Sheet Analysis:

--- OVER_DRAWL_wrt_CD ---
- Distribution of Total Overdrawn Count: Majority of consumers (13,191) had 0 overdrawn counts. Counts range from 1 to 5 for those who overdraw.
- Relationship between Contracted Load and Total Overdrawn Count: Consumers who overdraw tend to have a lower average Contracted Load compared to those who don't.
- Geographical Hotspots for Overdrawing: RANCHI, JAMSHEDPUR, and GIRIDIH zones; RANCHI, JAMSHEDPUR, and DEOGHAR circles; GODDA, RANCHI WEST, and KODERMA divisions showed the highest total overdrawn counts.
- Overall Overdrawing Proportion: 25.65% of consumers identified as overdrawing.
- Average contracted load: 20.07
- Contracted load range: 1.00 - 137.00
- Average MDI (Maximum Demand Indicator): 12.30
- Maximum demand (MDI value): 633.00
- Total Number of Consumers: 17742
- Total overdrawal occurrences (all months, all customers): 13837

--- MDI>=105%CD_4_MONTHS ---
- Distribution of EXCEEDANCE_COUNT: Many

## Summary:

### Data Analysis Key Findings

*   **OVER\_DRAWL\_wrt\_CD:**
    *   13,191 consumers had 0 overdrawn counts; counts ranged from 1 to 5 for those who overdraw.
    *   Consumers who overdraw tend to have a lower average Contracted Load than those who don't.
    *   RANCHI, JAMSHEDPUR, and GIRIDIH zones; RANCHI, JAMSHEDPUR, and DEOGHAR circles; and GODDA, RANCHI WEST, and KODERMA divisions showed the highest total overdrawn counts.
    *   25.65% of consumers were identified as overdrawing.
*   **MDI>=105%CD\_4\_MONTHS:**
    *   Many consumers had 0 exceedances, while a portion had 1 to 5 exceedances.
    *   Higher exceedance counts were associated with slightly lower average Contracted Load but higher average Enhanced Contract Demand.
    *   Top Zones, Circles, and Divisions with the highest sum of EXCEEDANCE\_COUNT were identified.
*   **MDI>=110%CD\_4\_MONTHS:**
    *   All consumers in this sheet had an excess demand count of 4 or 5.
    *   Consumers with a count of 5 had slightly lower average Contracted Load but higher average Enhanced Contract Demand.
    *   Top geographical areas with the highest sum of Excess\_Demand\_Count were identified.
    *   100% of consumers in this sheet met the specified condition.
*   **MDI=0\_ALL\_MONTHS\_UNITS>=100:**
    *   The majority of consumers had billed units above 100 ('YES').
    *   Consumers with 'YES' had a significantly higher average Contracted Load.
    *   Top geographical areas for consumers with '100units\_above' as YES were identified.
    *   Varied ranges and averages of consumption were observed for consumers with '100units\_above' as 'YES'.
*   **MDI=0\_LAST\_MONTH\_UNITS.=0:**
    *   All consumers had an MDI of 0 in the last month.
    *   The minimum billed units in the last month was 100, which contradicted the sheet name.
    *   A positive correlation between Contracted Load and last month's billed units was observed.
    *   Top geographical areas with the lowest average billed units in the last month were identified.
*   **SUMMARY:**
    *   The 'SUMMARY' sheet was empty and contained no data for analysis.

### Insights or Next Steps

*   Investigate the discrepancy between the sheet name 'MDI=0\_LAST\_MONTH\_UNITS.=0' and the actual minimum billed units of 100 in that sheet to understand the intended data.
*   Focus on the identified geographical hotspots for overdrawing and exceedances in the 'OVER\_DRAWL\_wrt\_CD', 'MDI>=105%CD\_4\_MONTHS', and 'MDI>=110%CD\_4\_MONTHS' sheets for targeted interventions or further analysis.


In [30]:
display(df_selected_sheet.describe())

S.NO.    ACCOUNT_NO  CONTRACTED_LOAD      MF_KWH  \
count  990.000000  9.900000e+02       990.000000  990.000000   
mean   495.500000  4.362307e+10        22.099525    1.019192   
std    285.932684  2.001839e+10        22.074766    0.603860   
min      1.000000  1.122072e+10         2.000000    1.000000   
25%    248.250000  3.122239e+10         8.000000    1.000000   
50%    495.500000  4.123599e+10        10.000000    1.000000   
75%    742.750000  6.122092e+10        26.000000    1.000000   
max    990.000000  7.225023e+10       107.000000   20.000000   

       BILLED_UNITS_HTS_202504  MDI_HTS_202504  BILLED_UNITS_HTS_202505  \
count               918.000000      918.000000             8.950000e+02   
mean               2176.044662        7.623222             6.341947e+03   
std                8212.505684       14.466207             1.097834e+05   
min                   0.000000        0.000000             1.000000e+00   
25%                 222.250000        0.000000             2.425000e+02   
50%                 635.000000        2.000000             6.820000e+02   
75%                1673.250000        8.000000             1.795500e+03   
max              178257.000000      103.000000             3.235850e+06   

       MDI_HTS_202505  BILLED_UNITS_HTS_202506  MDI_HTS_202506  \
count      895.000000               906.000000      906.000000   
mean         6.825942              2150.564018        5.620874   
std         13.971257             13432.126725       12.287890   
min          0.000000                 1.000000        0.000000   
25%          0.000000               237.250000        0.000000   
50%          1.000000               630.500000        0.000000   
75%          7.000000              1645.750000        6.000000   
max        114.000000            393292.000000       95.000000   

       LOAD_FACTOR_HTS_202506  SUPPLY_HOURS_HTS_202506  \
count                   906.0                    906.0   
mean                      0.0                      0.0   
std                       0.0                      0.0   
min                       0.0                      0.0   
25%                       0.0                      0.0   
50%                       0.0                      0.0   
75%                       0.0                      0.0   
max                       0.0                      0.0   

       BILLED_UNITS_HTS_202507  MDI_HTS_202507  LOAD_FACTOR_HTS_202507  \
count               926.000000      926.000000                   926.0   
mean               1936.917927        4.212527                     0.0   
std                7454.351558       10.425834                     0.0   
min                   1.000000        0.000000                     0.0   
25%                 239.000000        0.000000                     0.0   
50%                 694.500000        0.000000                     0.0   
75%                1750.250000        4.315000                     0.0   
max              203628.000000      109.000000                     0.0   

       SUPPLY_HOURS_HTS_202507  BILLED_UNITS_HTS_202508  MDI_HTS_202508  \
count                    926.0               990.000000           990.0   
mean                       0.0              2406.583838             0.0   
std                        0.0              9946.291281             0.0   
min                        0.0               100.000000             0.0   
25%                        0.0               260.000000             0.0   
50%                        0.0               763.000000             0.0   
75%                        0.0              1797.500000             0.0   
max                        0.0            207272.000000             0.0   

       LOAD_FACTOR_HTS_202508  SUPPLY_HOURS_HTS_202508  
count                   990.0                    990.0  
mean                      0.0                      0.0  
std                       0.0                      0.0  
min                       0.0                      0.0  
25%          